Fix crash while quitting in mpv_render_context_update, #5023

low-batt opened the pull request:
This commit will:
- Change where `isUninited` is locked in `VideoLayer`
- Add a `displayLock` property to `VideoLayer`
- Use `displayLock` to single thread the display method
- Move the skip render code from the draw method to the display method

This ensures the render context has not been freed when `shouldRenderUpdateFrame` is called. This also changes the location of the skip render code to match how mpv handles skipping.

- [x] I have read [CONTRIBUTING.md](https://github.com/iina/iina/blob/develop/CONTRIBUTING.md)
- [x] This implements/fixes issue #5023.

---

**Description:**

low-batt committed:
Fix crash while quitting in mpv_render_context_update, #5023

This commit will:
- Change where isUninited is locked in VideoLayer
- Add a displayLock property to VideoLayer
- Use displayLock to single thread the display method
- Move the skip render code from the draw method to the display method

This ensures the render context has not been freed when
shouldRenderUpdateFrame is called. This also changes the location of the
skip render code to match how mpv handles skipping.

low-batt commented:
I am troubled that I can not explain why trying to match the way mpv handles the layer locking resulted in a deadlock when used in IINA. The deadlock occurred then I used a `NSLock` for `displayLock`, which is what mpv does. Seems the call to `CATransaction.flush` in the `display` method ends up calling the `display` method again. So I was forced to use a `NSRecursiveLock`. Anyone have any guesses as to how mpv is able to just use a `NSLock`? @svobs, @krackers, thoughts?
<details>
<summary><b>Deadlock:</b></summary>

```text
Call graph:
    8260 Thread_836017   DispatchQueue_1: com.apple.main-thread  (serial)
    + 8260 start  (in dyld) + 2236  [0x195babf28]
    +   8260 main  (in IINA) + 12  [0x102d4daec]  AppDelegate.swift:24
    +     8260 NSApplicationMain  (in AppKit) + 880  [0x1991d3358]
    +       8260 -[NSApplication run]  (in AppKit) + 500  [0x1991fbf2c]
    +         8260 -[NSApplication _handleEvent:]  (in AppKit) + 60  [0x199584bd4]
    +           8260 -[NSApplication(NSEvent) sendEvent:]  (in AppKit) + 1556  [0x199334abc]
    +             8260 -[NSWindow(NSEventRouting) sendEvent:]  (in AppKit) + 284  [0x199335778]
    +               8260 -[NSWindow(NSEventRouting) _reallySendEvent:isDelayedEvent:]  (in AppKit) + 364  [0x199335ab8]
    +                 8260 -[NSWindow(NSEventRouting) _handleMouseDownEvent:isDelayedEvent:]  (in AppKit) + 3476  [0x1993aae7c]
    +                   8260 @objc InitialWindowViewActionButton.mouseDown(with:)  (in IINA) + 52  [0x102b1e0cc]  <compiler-generated>:0
    +                     8260 InitialWindowViewActionButton.mouseDown(with:)  (in IINA) + 2412  [0x102b1e01c]  InitialWindowController.swift:470
    +                       8260 PlayerCore.openURL(_:shouldAutoLoad:)  (in IINA) + 164  [0x102f7f598]  PlayerCore.swift:355
    +                         8260 PlayerCore.openURLs(_:shouldAutoLoad:)  (in IINA) + 1500  [0x102f7efe8]  PlayerCore.swift:339
    +                           8260 PlayerCore.open(_:shouldAutoLoad:)  (in IINA) + 1408  [0x102f7e988]  PlayerCore.swift:296
    +                             8260 PlayerCore.openMainWindow(path:url:isNetwork:)  (in IINA) + 1380  [0x102f808b4]  PlayerCore.swift:401
    +                               8260 -[NSWindowController window]  (in AppKit) + 84  [0x199372098]
    +                                 8260 -[NSWindowController _windowDidLoad]  (in AppKit) + 620  [0x199375ca4]
    +                                   8260 @objc MainWindowController.windowDidLoad()  (in IINA) + 36  [0x102c76790]  <compiler-generated>:0
    +                                     8260 MainWindowController.windowDidLoad()  (in IINA) + 4956  [0x102c74564]  MainWindowController.swift:573
    +                                       8260 PlayerCore.initVideo()  (in IINA) + 320  [0x102f84ad4]  PlayerCore.swift:508
    +                                         8260 @objc ViewLayer.display()  (in IINA) + 36  [0x102ebfb18]  <compiler-generated>:0
    +                                           8260 ViewLayer.display()  (in IINA) + 204  [0x102ebe9a8]  ViewLayer.swift:174
    +                                             8260 CA::Transaction::commit()  (in QuartzCore) + 648  [0x19d4cc2c0]
    +                                               8260 CA::Context::commit_transaction(CA::Transaction*, double, double*)  (in QuartzCore) + 456  [0x19d65d4d4]
    +                                                 8260 CA::Layer::display_if_needed(CA::Transaction*)  (in QuartzCore) + 760  [0x19d4e9ae8]
    +                                                   8260 @objc ViewLayer.display()  (in IINA) + 36  [0x102ebfb18]  <compiler-generated>:0
    +                                                     8260 ViewLayer.display()  (in IINA) + 116  [0x102ebe950]  ViewLayer.swift:170
    +                                                       8260 _pthread_mutex_firstfit_lock_slow  (in libsystem_pthread.dylib) + 248  [0x1074f3c98]
    +                                                         8260 _pthread_mutex_firstfit_lock_wait  (in libsystem_pthread.dylib) + 96  [0x1074f3d6c]
    +                                                           8260 __psynch_mutexwait  (in libsystem_kernel.dylib) + 8  [0x195ec6be8]
```
</details>

svobs commented:
In the current HEAD of `develop`, I see this comment above `CATransaction.flush()`:
https://github.com/iina/iina/blob/172e091776e9f02815aa5402c1366825c8965f5c/iina/ViewLayer.swift#L195-L197

But in this patch I see this:
https://github.com/iina/iina/blob/b4031123f23ec8eda357af2a6179192fcace4e9f/iina/ViewLayer.swift#L170-L175

If the original comment is correct, then `display` is being called while `displayLock` is still held. Perhaps `CATransaction.flush` is getting compiled into `CA::Transaction::commit`.

I question whether introducing another lock which overlaps the `isUninited` lock is the best approach. Seems like it would make for much cleaner & easier to understand code to refactor so that rendering only ever done via `mpvGLQueue`. I believe @krackers was the first to recommend this.

low-batt commented:
Excellent question. This:
```swift
// Must not call flush while holding isUninited's lock as that method may call display and our 
// locks do not support recursion. 
````
Has to do with `isUninited` using an implementation that does not support a thread locking the same lock a second time. Doing so will deadlock. This is not a problem for `displayLock` because it is using `NSRecursiveLock` which permits a thread to lock a lock more than once.

This particular locking pattern was chosen to match up with how mpv does it. _Except_ mpv is using a plain `NSLock` for `displayLock`. It really bothers me that I can not explain why `display` is not being called recursively in mpv allowing use of `NSLock`.

At one point I tried to change rendering to always use the `mpvGLQueue`. Users reported problems that I was unable to reproduce. Had to back out those changes.

svobs commented:
Oh sorry, I misread your question.

I think it might be a special case. Based on my testing the deadlock only seems to occur when `display` is first called in `PlayerCore`:
```Swift
  func initVideo() {
    // init mpv render context.
    // The video layer must be displayed once to get the OpenGL context initialized.
    mainWindow.videoView.videoLayer.display()  // <-- here
```

I tried this code (https://github.com/iina/iina/commit/0ad796bd9f11cc83ae56a7a99dab302570a7f2d1) as a crude test and it allowed a regular `NSLock` to be used, which seems to support this idea.

low-batt commented:
There is a better way to create the OpenGL context. I will experiment with that and see if I can switch this to `NSLock`.

low-batt commented:
I confirmed your findings, but I'm still trying to figure out why the mpv code does not need to do this.

low-batt commented:
Marked this as a draft while I investigate further.

low-batt committed:
Fix crash while quitting in mpv_render_context_update, #5023

This commit will:
- Change where isUninited is locked in VideoLayer
- Add a displayLock property to VideoLayer
- Use displayLock to single thread the display method
- Move the skip render code from the draw method to the display method
- Construct the OpenGL context and pixel format objects when the layer is
  initialized
- Remove the call to videoLayer.display in PlayerCore.initVideo that was called
  to trigger the creation of the OpenGL context
- Remove the call to videoLayer.draw in MainWindowController.windowWillOpen
  that was called to display a black window

The change in the location of locking ensures the render context has not been
freed when shouldRenderUpdateFrame is called. The change in location of the
skip render code was made to match how mpv handles skipping. The two calls must
be removed because they result in the call to CATransaction.flush recursively
calling display, which is not permitted by the type of lock being used.

low-batt commented:
With regards to the question of why mpv does not need to use a recursive lock and yet IINA deadlocks unless a recursive lock is used…

To answer that we have to guess why `CATransaction.flush()` sometimes calls `ViewLayer.display`:
```text
 8260 PlayerCore.initVideo()  (in IINA) + 320  [0x102f84ad4]  PlayerCore.swift:508
  8260 @objc ViewLayer.display()  (in IINA) + 36  [0x102ebfb18]  <compiler-generated>:0
    8260 ViewLayer.display()  (in IINA) + 204  [0x102ebe9a8]  ViewLayer.swift:174
      8260 CA::Transaction::commit()  (in QuartzCore) + 648  [0x19d4cc2c0]
        8260 CA::Context::commit_transaction(CA::Transaction*, double, double*)  (in QuartzCore) + 456  [0x19d65d4d4]
          8260 CA::Layer::display_if_needed(CA::Transaction*)  (in QuartzCore) + 760  [0x19d4e9ae8]
            8260 @objc ViewLayer.display()  (in IINA) + 36  [0x102ebfb18]  <compiler-generated>:0
              8260 ViewLayer.display()  (in IINA) + 116  [0x102ebe950]  ViewLayer.swift:170
```

The stack trace indicates `CA::Layer::display_if_needed` decided `display` needed to be called. Debugging shows [needsDisplay](https://developer.apple.com/documentation/appkit/nsview/1483360-needsdisplay) is `false` before calling [flush](https://developer.apple.com/documentation/quartzcore/catransaction/1448270-flush), however [needsLayout](https://developer.apple.com/documentation/appkit/nsview/1526912-needslayout) is `true`. It _appears_ that `display_if_needed` is calling `display` because layout is needed.

The code identified by @svobs shows that `PlayerCore.initVideo` is calling `display` only to initialize the OpenGL context:
```swift
    // The video layer must be displayed once to get the OpenGL context initialized.
    mainWindow.videoView.videoLayer.display()
    mpv.mpvInitRendering()
```

To fix this the latest commit forms the OpenGL context (and pixel format) when the `ViewLayer` object is constructed. With that change the call to `display` in `initVideo` can be removed.

However fixing that uncovered a second call that may occur before laytout. This one is in `MainWindowController.windowWillOpen`:
```swift
    videoView.videoLayer.draw(forced: true)
```
This call was added in commit f921f4a47dc8da437e414c8c77daa08287b39287 with the comment:
> Draw black screen before playing a video

I discussed this with @uiryuu and he indicated this call can be removed.

With these changes I was able to use IINA's `Lock` class instead of a recursive lock.

low-batt commented:
I did some more testing and managed to find another way to trigger a deadlock. Investigating…

low-batt committed:
Fix crash while quitting in mpv_render_context_update, #5023

This commit will:
- Change where isUninited is locked in VideoLayer
- Add a displayLock property to VideoLayer
- Use displayLock to single thread the display method
- Move the skip render code from the draw method to the display method
- Construct the OpenGL context and pixel format objects when the layer is
  initialized
- Remove the call to videoLayer.display in PlayerCore.initVideo that was called
  to trigger the creation of the OpenGL context

The change in the location of locking ensures the render context has not been
freed when shouldRenderUpdateFrame is called. The change in location of the
skip render code was made to match how mpv handles skipping.

low-batt commented:
The new deadlock I hit was with switching out of music mode. Given the behavior of  `CATransaction.flush` I think it is risky to not use a recursive lock. So the latest commit switches this back to `NSRecursiveLock`.

uiryuu commented on iina/ViewLayer.swift:
```diff
@@ -43,92 +54,62 @@ class ViewLayer: CAOpenGLLayer {
+                        forLayerTime t: CFTimeInterval, displayTime ts: UnsafePointer<CVTimeStamp>?) -> Bool {
+    videoView.$isUninited.withLock() { isUninited in
+      guard !isUninited else { return false }
+      guard !forceRender else { return true }
```
```suggestion
      if forceRender { return true }
```

low-batt replied:
Will fix.

uiryuu commented on iina/ViewLayer.swift:
```diff
@@ -17,12 +17,23 @@ class ViewLayer: CAOpenGLLayer {
+
   private var fbo: GLint = 1
 
   private var needsMPVRender = false
   private var forceRender = false
```
I think we should combine these two variables into a single variable to improve readibility, something like
```swift
private var needsRender = [.no, .yes. force]
```

low-batt replied:
We could, but that would deviate from how mpv handles this.

uiryuu commented on iina/ViewLayer.swift:
```diff
@@ -154,10 +135,18 @@ class ViewLayer: CAOpenGLLayer {
+    defer { displayLock.unlock() }
+
+    super.display()
+    CATransaction.flush()
```
Why we need to flush here?

low-batt replied:
Implicit transactions that are not on the main thread must be manually flushed. On the main thread they are automatically flushed at the end of the run loop. There is a question as to whether we should be creating implicit transactions off the main thread. See issue #5038.

uiryuu commented:
> To fix this the latest commit forms the OpenGL context (and pixel format) when the ViewLayer object is constructed. With that change the call to display in initVideo can be removed.

I remember because of this, when we open a window, we must draw a black window first in order to make the GLLayer init correctly; at that point, mpv isn't ready, so we don't know the size of the video, which causes the window being stretched to the final window size. With this fix, I think we can finally display the window after the size is available, thus avoiding the arguably ugly stretch animation (which some users has complained about).

svobs commented:
> I did some more testing and managed to find another way to trigger a deadlock. Investigating…

Can you elaborate on this? It sounds like you are saying that doing a "forced" draw can trigger a deadlock? Or is the deadlock happening each time a `layout` is done?

It would be great to identify all the places the video layer's `display` method is being called, because they should be very limited. I recently found documentation on [layerContentsRedrawPolicy](https://developer.apple.com/documentation/appkit/nsview/1483514-layercontentsredrawpolicy) which states the following:

> For a view that has no associated layer, or that has been assigned a developer-provided layer (a layer-hosting view) using the [layer](https://developer.apple.com/documentation/appkit/nsview/1483298-layer) property, the default contents redraw policy is [NSView.LayerContentsRedrawPolicy.never](https://developer.apple.com/documentation/appkit/nsview/layercontentsredrawpolicy/never) and the [layerContentsPlacement](https://developer.apple.com/documentation/appkit/nsview/1483375-layercontentsplacement) property is set to [NSView.LayerContentsPlacement.scaleAxesIndependently](https://developer.apple.com/documentation/appkit/nsview/layercontentsplacement/scaleaxesindependently). These policies tell AppKit not to replace the layer’s content and to provide the same content placement as the [resize](https://developer.apple.com/documentation/quartzcore/calayercontentsgravity/1410811-resize) option.

I confirmed via testing that `VideoView` has redraw policy = `never`. So it would follow that its `needsDisplay` will always be `false` as well (unless changed in our code). So what are the consequences of that? Well, I stumbled onto [this dense post](https://stackoverflow.com/a/11213382/1347529) which I am still digesting. It might also be interesting for reasons outside this PR.

lhc70000 commented:
> This call was added in commit https://github.com/iina/iina/commit/f921f4a47dc8da437e414c8c77daa08287b39287 with the comment:
> 
> Draw black screen before playing a video
> 
> I discussed this with @uiryuu and he indicated this call can be removed.

I remember that this was added to fix a rare issue: sometimes when opening a video, the last frame of the video played before will be displayed in the window before the playback starts. This happened more than once when the user complains IINA reveals some private video they played last time. We wasn't able to readily reproduce this, so we added a general fix.

IMO the most possible reason is that the newly created layer may contain random or undefined data, and we should manually clear it before use. So even this call is removed, we need to call `glClear()` somewhere after creating the layer and before displaying it.

low-batt commented:
When I went back to using a recursive lock I added the drawing back in to `windowDidOpen`. It bothered me that I could not explain why it was necessary to black out the screen, so I was worried about removing it.

That it has to do with privacy and window reuse explains why IINA differs from mpv in this case as it is unlikely mpv is reusing windows. This is also an example of code whose purpose is not immediately clear and could use a comment. I will add one.

I believe this means the problem @uiryuu mentioned [above](https://github.com/iina/iina/pull/5029#issuecomment-2205228734) with IINA drawing before the size of the video is known is still present and needs to be addressed.

On this:
> newly created layer may contain random or undefined data

I think it is more straight forward than that? The cached player contains the window, view and layer. The previous layer is being reused. I set a breakpoint in `ViewLayer.init` and it was only hit once when the window was initially created. Closing the window and opening another video did not create a new layer.

That should mean that instead of drawing a black screen at the start and thus causing the issue @uiryuu mentioned we should be able to clear the layer when the window is closing. This should be addressed in a separate PR.

krackers commented:
@svobs 

>So what are the consequences of that?

Fwiw option 2 in that post is what iina does currently in the non live-resize case, while option 1 is what it (now) does during live-resize. The `setNeedsDisplayOnBoundsChange:YES` is interesting, I'm not sure if setting this would solve the blurriness during live-resize.

Also the line
> Internally a NSOpenGLView creates a NSOpenGLLayer when you make it layered,

is no longer true, on 10.14+ (?) a layer-backed NSOpenGLView differs from a NSOpenGLLayer. In fact even if you don't declare an NSOpenGLView as explicitly layer-backed, it still appears to composite to a layer. This is the only way to get push based semantics for opengl drawing with layer-backing.



---

>newly created layer may contain random or undefined data, and we should manually clear it before use

I've seen this happen with raw nsopenglview on older os (which makes sense since an opengl surface is basically mapped almost directly onto a chunk of vram), but I would have thought that a layer-backed view would properly handle this to avoid compositing garbage. Maybe you could set `backgroundColor` on the layer? 

Edit: Missed lowbatt's comment, if it's indeed the same layer being reused (and not a new layer) that would explain it.

low-batt commented:
On the question from @svobs about when the deadlock occurs…

I don't think this Apple code is open source? At least I have not been able to find it. So we have to guess what is going on.

From that stack trace we know the call to `CATransaction.flush` causes the implicit transaction to be committed. As a part of that `display_if_needed` is called and that method decides `display` should be called. This we know for certain.

As mpv is not using a recursive lock apparently how they are drawing never results in `display_if_needed` calling `display`. Seems something IINA is doing affects the behavior of `flush`.

From logging the state of `needsLayout` before the call to `CATransaction.flush` it appears there is a correlation between needing layout and `display_if_needed` calling `display`. The cases I found so far where this happened was the early calls to display when starting to play a video and switching from music mode to the main window. These are all operations mpv does not need to do.

@saagarjha is attempting to confirm this is expected `CATransaction.flush` behavior.

lhc70000 commented:
> I think it is more straight forward than that? The cached player contains the window, view and layer. The previous layer is being reused. I set a breakpoint in ViewLayer.init and it was only hit once when the window was initially created. Closing the window and opening another video did not create a new layer.

Yes, that might also be a reason. However I vaguely remember that this could also happen when you open a fresh IINA instance. Of course it was a very old issue, and things may have changed now.

low-batt committed:
Fix crash while quitting in mpv_render_context_update, #5023

This commit will:
- Change where isUninited is locked in VideoLayer
- Add a displayLock property to VideoLayer
- Use displayLock to single thread the display method
- Move the skip render code from the draw method to the display method
- Construct the OpenGL context and pixel format objects when the layer is
  initialized
- Remove the call to videoLayer.display in PlayerCore.initVideo that was called
  to trigger the creation of the OpenGL context

The change in the location of locking ensures the render context has not been
freed when shouldRenderUpdateFrame is called. The change in location of the
skip render code was made to match how mpv handles skipping.

low-batt commented:
For a fresh IINA process to display a frame from a previously played video likely means macOS had a security defect. The OS can cache and reuse some things, such as file buffers. But a new process should not have any access to graphics memory from some other process.

Maybe someone thought IINA was being opened fresh when it was still open in the dock, just with no windows currently open?

svobs commented:
> The cases I found so far where this happened was the early calls to display when starting to play a video and switching from music mode to the main window.

Did some more testing to bring out the deadlock. I reproduced it via the "forced" display while exiting music mode, but not entering... Oddly, I was not able to make it deadlock when toggling PiP, which should be very similar to the music mode toggle. Possibly a race condition? Was hoping to find more of a pattern, but doubt it's crucial to keep dwelling on it if the recursive lock fixes it.
